### ISSUE TITLE
(Window | Worker).addEventListener(...)

### ISSUE THREAD

**Issue by ScottSheehan** [microsoft/TypeScript#48546]
# Bug Report

Adding a "message" event listener to a var that is type `Worker | Window` results in a type error. However if you cast the variable to either Window or Worker (individually), it does not produce a type error (for either).

### 🔎 Search Terms
Argument of type '(e: MessageEvent<any>)

    Argument of type '(e: MessageEvent<any>) => void' is not assignable to parameter of type 'EventListenerOrEventListenerObject'.
  Type '(e: MessageEvent<any>) => void' is not assignable to type 'EventListener'.
    Types of parameters 'e' and 'evt' are incompatible.
      Type 'Event' is missing the following properties from type 'MessageEvent<any>': data, lastEventId, origin, ports, and 2 more.

### 🕗 Version & Regression Information

On bug-workbench, reproes in 3.3.3 through nightly (4.7)

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/DYUwLgBAZhC80FcB2BjMBLA9kgFCAXBALIgDOpAhgOYgCiAbiEmAJQQDeAvgLABQfUZGixIIYHAHdCAdUwAnANYg5EAD4Rp6JABNMEtuz4RjESRAqkNW3foB0FbdoZMwAGXSkwTZTgBEAWzJKGl8AGmgWAG4IAHoYqC0QIxMzCw15JTkWe0dnZndPbzk-QPJqEDCI6LiKYFJMaETk4wkcp0Z8jy8kHwCg8sqoKNiYiGU5eXxm2IAqaYBBOSoEQOYITBgwAE8ABxAIAHI8QhIymjywAB4KJC2APjZYO4h6THRtA4gPCCRMSAtSOgqEgKAAjUBiBo7ChyCiBLwqDZiXb7A4XArdZQAeTk6K6RSxoIAViA0AdbMkACoow7HYj9c4dK43e6PZ6vd6fb6-f7kIEg8H7MANbZ7Q54wo9OTk6bUvaWJHQ2Hw5SWA4gT43bSHED0MCauT7LQoTD+aEYQUU3gmExy1EXLmWfweQFIKhiAAW+ygmGAwD0WndOwmezkGDI0Am-mRYoOp2CdCZ11udwOhG0FDAFHCwAsYAuAEltOF5ECtOEdvIwKRwlqIAAmCD+eQgK02iAzGJ8ThAA)

### 💻 Code

```ts
let f = function(e: MessageEvent) {}

function t(w: Worker | Window) {
    (w as Window).addEventListener("message", f); //fine
    (w as Worker).addEventListener("message", f); //also fine
    w.addEventListener("message", f); // error:
    /*
    Argument of type '(e: MessageEvent<any>) => void' is not assignable to parameter of type 'EventListenerOrEventListenerObject'.
  Type '(e: MessageEvent<any>) => void' is not assignable to type 'EventListener'.
    Types of parameters 'e' and 'evt' are incompatible.
      Type 'Event' is missing the following properties from type 'MessageEvent<any>': data, lastEventId, origin, ports, and 2 more.
     */
}
```

### 🙁 Actual behavior

Type error.

### 🙂 Expected behavior

No type error.

**Comment by MartinJohns** [comment]
This could be a duplicate of #33101, or at least be related. Found by searching for `union overloads in:title`.

The `addEventListener` method of your union type does not have the overloads accepting specific keys anymore, so it falls back to the generic `string` version. That one requires a more specific type than `MessageEvent`.

This is the simplified case of your issue:

```typescript
type A = {
  (kind: 'a' | 'b', value: boolean): void;
  (kind: string, value: string): void;
}

type B = {
  (kind: 'b' | 'c', value: boolean): void;
  (kind: string, value: string): void;
}


declare const a: A;
declare const b: B;
declare const ab: A | B;

a('b', true);
b('b', true);
ab('b', true);
```

One could assume that `A | B` should get reduced to:

```typescript
{
  (kind: 'b', value: boolean): void;
  (kind: string, value: string): void;
}
```

But it ends up as:

```typescript
{
  (kind: string, value: string): void;
}
```

**Comment by RyanCavanaugh** [comment]
We don't have any mechanism to do the sort of "deep" signature merging as described above; it would likely be impractical in terms of performance - you'd really have to do an n^2 cross-product to see if any signature could merge with any other signature, and might hit order dependence once there's more than one type in the union.
